extension/observer/kafkatopicsobserver/extension.go

```diff
@@ -19,113 +18,109 @@ import (
 )
 
 var (
-	_ extension.Extension      = (*kafkaTopicsObserver)(nil)
-	_ observer.EndpointsLister = (*kafkaTopicsObserver)(nil)
-	_ observer.Observable      = (*kafkaTopicsObserver)(nil)
+	_ extension.Extension = (*kafkaTopicsObserver)(nil)
+	_ observer.Observable = (*kafkaTopicsObserver)(nil)
 )
 
 type kafkaTopicsObserver struct {
 	*observer.EndpointsWatcher
 	logger           *zap.Logger
 	config           *Config
-	doneChan         chan struct{}
 	cancelKafkaAdmin func()
-	once             *sync.Once
-	topics           []string
-	kafkaAdmin       sarama.ClusterAdmin
-	mu               sync.Mutex
+	adminClient      sarama.ClusterAdmin
 }
 
 func newObserver(logger *zap.Logger, config *Config) (extension.Extension, error) {
-	kCtx, cancel := context.WithCancel(context.Background())
+	topicRegexp, err := regexp.Compile(config.TopicRegex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile topic regex: %w", err)
+	}
 
-	admin, err := createKafkaClusterAdmin(kCtx, *config)
+	kCtx, cancel := context.WithCancel(context.Background())
+	adminClient, err := createKafkaClusterAdmin(kCtx, *config)
 	if err != nil {
 		cancel()
 		return nil, fmt.Errorf("could not create kafka cluster admin: %w", err)
 	}
 
-	d := &kafkaTopicsObserver{
+	o := &kafkaTopicsObserver{
 		logger:           logger,
 		config:           config,
-		once:             &sync.Once{},
-		topics:           []string{},
 		cancelKafkaAdmin: cancel,
-		kafkaAdmin:       admin,
-		doneChan:         make(chan struct{}),
+		adminClient:      adminClient,
 	}
-	d.EndpointsWatcher = observer.NewEndpointsWatcher(d, time.Second, logger)
-	return d, nil
+	o.EndpointsWatcher = observer.NewEndpointsWatcher(
+		&kafkaTopicsEndpointsLister{o: o, topicRegexp: topicRegexp},
+		config.TopicsSyncInterval,
+		logger,
+	)
+	return o, nil
 }
 
-func (k *kafkaTopicsObserver) ListEndpoints() []observer.Endpoint {
-	// mutex is used in order to test the extensions without "race detected during execution of test" error
-	k.mu.Lock()
-	defer k.mu.Unlock()
-	endpoints := make([]observer.Endpoint, 0, len(k.topics))
-	for _, topic := range k.topics {
+func (k *kafkaTopicsObserver) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+func (k *kafkaTopicsObserver) Shutdown(_ context.Context) error {
+	k.StopListAndWatch()
+	k.cancelKafkaAdmin()
+	err := k.adminClient.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close kafka cluster admin client: %w", err)
+	}
+	k.logger.Info("kafka cluster admin client closed")
+	return nil
+}
+
+type kafkaTopicsEndpointsLister struct {
+	o           *kafkaTopicsObserver
+	topicRegexp *regexp.Regexp
+
+	mu     sync.Mutex
+	topics []string
+}
```

Comment:
Any reason to move this to own type instead of the original combined?

Reply:
Basically the same answer as https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38412#discussion_r1982650472, just API hygiene -- the EndpointsLister does not need to be exposed to external users of the observer.